fe/fe-core/src/main/java/com/starrocks/sql/optimizer/InputPropertyDeriver.java

```diff
@@ -0,0 +1,290 @@
+
+        PhysicalPropertySet leftInputProperty = createPropertySetByDistribution(leftDistribution);
+        PhysicalPropertySet rightInputProperty = createPropertySetByDistribution(rightDistribution);
+
+        inputProperties.add(Lists.newArrayList(leftInputProperty, rightInputProperty));
+        return null;
+    }
+
+    private Optional<HashDistributionDesc> getRequiredLocalDesc() {
+        if (!requirements.getDistributionProperty().isShuffle()) {
+            return Optional.empty();
+        }
+
+        HashDistributionDesc requireDistributionDesc =
+                ((HashDistributionSpec) requirements.getDistributionProperty().getSpec()).getHashDistributionDesc();
+        if (!HashDistributionDesc.SourceType.SHUFFLE_LOCAL.equals(requireDistributionDesc.getSourceType())) {
+            return Optional.empty();
+        }
+
+        return Optional.of(requireDistributionDesc);
+    }
+
+    @Override
+    public Void visitPhysicalHashAggregate(PhysicalHashAggregateOperator node, ExpressionContext context) {
+        // If scan tablet sum leas than 1, do one phase local aggregate is enough
+        if (ConnectContext.get().getSessionVariable().getNewPlannerAggStage() == 0
+                && context.getRootProperty().isExecuteInOneTablet()
+                && node.getType().isGlobal() && !node.isSplit()) {
+            inputProperties.add(Lists.newArrayList(PhysicalPropertySet.EMPTY));
+            return null;
+        }
+
+        LogicalOperator child = (LogicalOperator) context.getChildOperator(0);
+        // If child has limit, we need to gather data to one instance
+        if (child.hasLimit() && (node.getType().isGlobal() && !node.isSplit())) {
+            inputProperties.add(Lists.newArrayList(createLimitGatherProperty(child.getLimit())));
+            return null;
+        }
+
+        if (!node.getType().isLocal()) {
+            List<Integer> columns = node.getPartitionByColumns().stream().map(ColumnRefOperator::getId).collect(
+                    Collectors.toList());
+
+            // None grouping columns
+            if (columns.isEmpty()) {
+                DistributionProperty distributionProperty =
+                        new DistributionProperty(DistributionSpec.createGatherDistributionSpec());
+                inputProperties.add(Lists.newArrayList(new PhysicalPropertySet(distributionProperty)));
+                return null;
+            }
+
+            // shuffle aggregation
+            DistributionSpec distributionSpec = DistributionSpec.createHashDistributionSpec(
+                    new HashDistributionDesc(columns, HashDistributionDesc.SourceType.SHUFFLE_AGG));
+            DistributionProperty distributionProperty = new DistributionProperty(distributionSpec);
+            inputProperties.add(Lists.newArrayList(new PhysicalPropertySet(distributionProperty)));
+            return null;
+        }
```

Comment:
I suggest refactor this code，like:
```
if (node.isOneStage()) {
......
}
if (node.isTwoStage()) {
......
}
if (node.isThreeStage()) {
......
}
......
```

or give some comment to describe the propery for every stage

Reply:
we can not figure out the node belonges to what stage here